app/Cardano/CLI.hs

```diff
@@ -12,103 +11,145 @@
-    encode Local = "local"
-
-instance Decodable String Network where
-    decode "mainnet" = Right Mainnet
-    decode "testnet" = Right Testnet
-    decode "staging" = Right Staging
-    decode "local" = Right Local
-    decode s = Left $ show s ++
-        " is neither \"mainnet\", \"testnet\", \"staging\" nor \"local\"."
-
-instance Decodable [String] (Mnemonic 15) where
-    decode ws = first show $ mkMnemonic @15 (T.pack <$> ws)
+-- | Port number with a tag for describing what it is used for
+newtype Port (tag :: Symbol) = Port Int
+
+instance FromText (Port tag) where
+    fromText = fmap Port . fromText
+
+instance ToText (Port tag) where
+    toText (Port p) = toText p
+
+{-------------------------------------------------------------------------------
+                             Parsing Arguments
+-------------------------------------------------------------------------------}
+
+parseArgWith :: FromText a => Docopt -> Arguments -> Option -> IO a
+parseArgWith cli args option =
+    either (fail . getTextDecodingError) pure . fromText . T.pack
+        =<< args `getArgOrExit` option
+  where
+    getArgOrExit :: Arguments -> Option -> IO String
+    getArgOrExit = getArgOrExitWith cli
+
+{-------------------------------------------------------------------------------
+                         Processing of Sensitive Data
+-------------------------------------------------------------------------------}
+
+-- | Repeatedly prompt a user for a sensitive value, until the supplied value is
+-- valid.
+getRequiredSensitiveValue
+    :: Buildable e
+    => (Text -> Either e a)
+    -> String
+    -> IO a
+getRequiredSensitiveValue parse prompt = loop where
+    loop = do
+        putStrLn prompt
+        line <- getLineWithSensitiveData
+        case parse line of
+            Left e -> do
+                TIO.putStrLn (pretty e)
+                loop
+            Right v -> pure v
+
+-- | Repeatedly prompt a user for an optional sensitive value, until either the
+-- supplied value is valid, or until the user enters an empty line (indicating
+-- that they do not wish to specify such a value).
+getOptionalSensitiveValue
+    :: Buildable e
+    => (Text -> Either e a )
+    -> String
+    -> IO (Maybe a)
+getOptionalSensitiveValue parse prompt = loop where
+    loop = do
+        putStrLn prompt
+        line <- getLineWithSensitiveData
+        if T.length line == 0
+        then pure Nothing
+        else case parse line of
+            Left e -> do
+                TIO.putStrLn (pretty e)
+                loop
+            Right v ->
+                pure $ Just v
+
+-- | Read a line of user input containing sensitive data from the terminal.
+--
+-- The terminal lines containing the data are cleared once the user has finished
+-- entering data.
+--
+-- The terminal lines containing the data are also cleared if the application
+-- exits abnormally, before the user has finished entering data.
+--
+getLineWithSensitiveData :: IO Text
```

Comment:
alternitavelly we could even hide user input from the screen (similar like `sudo` does). 

The cons of hiding the user input is that user is unable to check/verify/easily-correct entered mnemonic or password.

Some CLI application approach the problem of requesting the same input data to be done twice in order to confirm that user didn't make a mistake:

```
Passphrase: ********
Retype passphrase: ********
```
user experience is lower with this model but users are now able to enter their their mnemonics or passphrases in public spaces.

Mnemonics are pretty long (15+ words) and passphrase as well (10+ chars) so there is a high chance of mistake. That's also the case not to implement what I am proposing.

I guess CLI will most likely be used by someone already more technically advanced - so there is a higher chance that person already is aware that mnemonic and passphrase is something he should hide and not reveal (ie, type in public place).

Conclusion is that I think this is ok solution but maybe we should keep in mind and research alternatives.